### PR TITLE
Allow Z in datetime string parsing in non pandas compat mode

### PR DESCRIPTION
**Pull request by mroeschke**
## Description
xref https://github.com/rapidsai/cudf/issues/14039#issuecomment-1874672202

## Checklist
- [ ] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.

### CODE REVIEW

File: python/cudf/cudf/tests/test_datetime.py [hunk 1]
```diff
@@ -1306,8 +1306,11 @@ def test_datetime_infer_format(data, timezone, dtype):
 
         assert_eq(expected, actual)
     else:
-        with pytest.raises(NotImplementedError):
-            sr.astype(dtype)
+        with cudf.option_context("mode.pandas_compatible", True):
+            with pytest.raises(NotImplementedError):
+                sr.astype(dtype)
+
+        # pandas doesn't allow parsing "Z" to naive type
```
Comment (bdice) [inline python/cudf/cudf/tests/test_datetime.py:1313]: Please move this comment above the `sr.astype(dtype)` call to make it clear that it's the cause of the error.

File: python/cudf/cudf/tests/test_datetime.py [hunk 2]
```diff
@@ -2308,12 +2311,22 @@ def test_format_timezone_not_implemented(code):
         )
 
 
-@pytest.mark.parametrize("tz", ["Z", "UTC-3", "+01:00"])
-def test_no_format_timezone_not_implemented(tz):
+@pytest.mark.parametrize("tz", ["UTC-3", "+01:00"])
+def test_utcoffset_not_implemented(tz):
```
Comment (bdice) [inline python/cudf/cudf/tests/test_datetime.py:2315]:
```suggestion
def test_utc_offset_not_implemented(tz):
```

File: python/cudf/cudf/tests/test_datetime.py [hunk 3]
```diff
@@ -2308,12 +2311,22 @@ def test_format_timezone_not_implemented(code):
...
     with pytest.raises(NotImplementedError):
         cudf.to_datetime([f"2020-01-01 00:00:00{tz}"])
 
 
+def test_Z_utcoffset():
```
Comment (bdice) [inline python/cudf/cudf/tests/test_datetime.py:2320]:
```suggestion
def test_Z_utc_offset():
```

**Review by bdice — COMMENTED** [review]
A couple small comments. We need to follow up with a deprecation for the behavior that doesn't align with pandas, and eventually make a breaking change for that since it's a bug that cudf and pandas disagree on the behavior. We'll need the Morpheus team (@cwharris) to update their code accordingly.

**Review by cwharris — APPROVED** [review]

File: python/cudf/cudf/tests/test_datetime.py (hunk 1 above)
Comment (mroeschke) [reply python/cudf/cudf/tests/test_datetime.py]: Sure thing. Moved

**Review by bdice — APPROVED** [review]

**Pull request merged**